test/transpile.test.js

```diff
@@ -14,35 +15,85 @@ function cleanCRLF(str) {
 }
 
 var transpileDir = path.join(__dirname, 'transpile');
-var expectedFilePath = path.join(transpileDir, 'expected.js');
-var expected = readCleanCRLF(expectedFilePath);
-
-/**
- * Execute the binary and capture the stdout
- */
-const { exec } = require('child_process');
-const child = exec('node ./bin/dogescript.js ./test/transpile/source.djs',
-  (err, stdout, stderr) => {
-    if(err)
-    {
-      console.error(`exec error: ${err}`);
-      return;
-    }
+
+// Pop a directory off it's path (os safe)
+function dotdotSlash(filePath) {
+  var str;
+  if (filePath.indexOf('/') == -1) { // windows
+    str = filePath.substring(0, filePath.lastIndexOf('\\'));
+  }
+  else { // unix
+    str = filePath.substring(0, filePath.lastIndexOf('/'));
   }
-);
 
-// stream the output
-var readable = child.stdout;
-readable.setEncoding('utf8');
+  return str + '/'
+}
 
-// when the process sends us data, validate they look the same
-readable.on('data', function(chunk) {
+function getFolderName(filePath) {
+  
+  // windows
+  if (filePath.indexOf('/') == -1) { 
+    return filePath.substring(filePath.lastIndexOf('\\') + 1, filePath.length);
+  }
+  
+  // unix
+  return filePath.substring(filePath.lastIndexOf('/') + 1, filePath.length);
+}
 
-  var actual = cleanCRLF(chunk.toString());
-  test('transpilation should work', function (t) {
-    t.plan(1);
+var testDirs = {};
 
-    t.equal(actual,expected);
-  });
+// Who's your daddy and what does he do?
+var skywalker = walk.walk(transpileDir);
+
+// Walk every folder in the file tree adding any folders that;
+// have an expected.js and have a source.js
+skywalker.on('names', function(fpath, children) {
+  var missingExpectFile = children.indexOf('expected.js') === -1;
+  var missingSourceFile = children.indexOf('source.djs') === -1;
+
+  
+  if(missingExpectFile || missingSourceFile) return;
+  
+  testDirs[fpath] = { missingExpectFile, missingSourceFile };
+});
 
+skywalker.on('end', function() {
+
+  var keys = Object.keys(testDirs);
+  keys.forEach(dir => {
+
+    var expectPath   = path.join(dir, 'expected.js');
+    var sourcePath     = path.join(dir, 'source.djs');
+    var expected = readCleanCRLF(expectPath);
+
+    /**
+     * Execute the binary and capture the stdout
+     */
+    const { exec } = require('child_process');
+    const child = exec(`node ./bin/dogescript.js ${sourcePath}`,
+      (err, stdout, stderr) => {
+        if(err)
+        {
+          console.error(`exec error: ${err}`);
+          return;
+        }
+      }
+    );
+
+    // stream the output
+    var readable = child.stdout;
+    readable.setEncoding('utf8');
+
+    // when the process sends us data, validate they look the same
+    readable.on('data', function(chunk) {
+
+      var actual = cleanCRLF(chunk.toString());
+      test(`transpilation of ${path.basename(dir)} should work`, function (t) {
```

Comment:
defining the test inside a callback seems dangerous

Reply:
I didn’t find a better way to guarantee that the transpiled js actually existed other than binding to the callback when the process finished (the process that pipes it to the dogescript binary)

Do you know of any other work arounds off the top of your head?

Reply:
K, i think i found a better way of doing this. Updated!